## Практическое задание: pandas + matplotlib

In [86]:
!pip install pandas matplotlib gdown

Defaulting to user installation because normal site-packages is not writeable


# Часть 1. Pandas

In [87]:
import pandas as pd

In [88]:
!gdown 1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6

^C


Downloading...
From: https://drive.google.com/uc?id=1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6
To: c:\Users\janep\Desktop\codes\57Sch\2024\Data Analysing\Combined10.10\youtube_trending_mediterranean.csv

  0%|          | 0.00/6.22M [00:00<?, ?B/s]
  8%|▊         | 524k/6.22M [00:00<00:03, 1.62MB/s]
 17%|█▋        | 1.05M/6.22M [00:00<00:02, 2.19MB/s]
 25%|██▌       | 1.57M/6.22M [00:00<00:01, 2.55MB/s]
 34%|███▎      | 2.10M/6.22M [00:00<00:01, 2.77MB/s]
 42%|████▏     | 2.62M/6.22M [00:01<00:01, 2.89MB/s]
 51%|█████     | 3.15M/6.22M [00:01<00:01, 2.98MB/s]
 59%|█████▉    | 3.67M/6.22M [00:01<00:00, 3.06MB/s]
 67%|██████▋   | 4.19M/6.22M [00:01<00:00, 3.03MB/s]
 76%|███████▌  | 4.72M/6.22M [00:01<00:00, 3.26MB/s]
 84%|████████▍ | 5.24M/6.22M [00:01<00:00, 3.24MB/s]
 93%|█████████▎| 5.77M/6.22M [00:01<00:00, 3.23MB/s]
100%|██████████| 6.22M/6.22M [00:02<00:00, 3.22MB/s]
100%|██████████| 6.22M/6.22M [00:02<00:00, 2.95MB/s]


#### Задание 1
* Откройте файл с таблицей, загруженной выше, и выведите последние 5 строк.
* Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк и столбцов?

In [89]:
# data = Your code here

df = pd.read_csv('youtube_trending_mediterranean.csv', sep = ',')
df.tail(5)

country     video_id                                              title  \
3216     NaN  eB-A1Y0Dswc                             Thank you for watching   
3217     NaN  KJGTtzccOHU  Pokemon Emerald but every battle is extremely ...   
3218     NaN  be8mqaTi_DI  Victor Cibrian x Fuerza Regida x Luis R Conriq...   
3219     NaN  14aYHc7ybu0  WE STAYED AT A 1 STAR HOTEL FOR 24 HOURS | coc...   
3220     NaN  mQ9eFI13NCQ  I 100%'d Cuphead: The Delicious Last Course, H...   

               publishedAt                 channelId  \
3216  2022-07-05T21:18:33Z  UCS5tt2z_DFvG7-39J3aE-bQ   
3217  2022-07-05T01:55:10Z  UC0VVYtw21rg2cokUystu2Dw   
3218  2022-07-04T21:00:36Z  UC648rgJOboZlgcDbW00vTSA   
3219  2022-07-05T13:40:04Z  UCSGoIq_tVESqNYF1Re-zn1Q   
3220  2022-07-04T16:04:33Z  UCRgzckrm1Svf-Y_dIWsTPTg   

                  channelTitle  categoryId trending_date  \
3216             Life of Boris          22      22.11.07   
3217                  SmallAnt          20      22.11.07   
3218            Rancho Humilde          10      22.11.07   
3219                 HJ Evelyn          24      22.11.07   
3220  The Andrew Collette Show          20      22.11.07   

                                                   tags  view_count  \
3216  lifeofboris|life of boris|english subtitles|ye...      864216   
3217  pokemon|smallant|smallant1|challenge|pokemon c...     1311698   
3218  Rancho Humilde|Legado 7|El De La Guitarra|Fuer...     2256933   
3219  hjevelyn|Evelyn ha|Emily ha|erica ha|korean si...      922668   
3220  cuphead|cuphead dlc|cuphead the delicious last...     1176289   

      comment_count                                  thumbnail_link  \
3216          14431  https://i.ytimg.com/vi/eB-A1Y0Dswc/default.jpg   
3217           1232  https://i.ytimg.com/vi/KJGTtzccOHU/default.jpg   
3218            597  https://i.ytimg.com/vi/be8mqaTi_DI/default.jpg   
3219           2951  https://i.ytimg.com/vi/14aYHc7ybu0/default.jpg   
3220            620  https://i.ytimg.com/vi/mQ9eFI13NCQ/default.jpg   

      comments_disabled  ratings_disabled  \
3216              False              True   
3217              False              True   
3218              False              True   
3219              False              True   
3220              False              True   

                                            description  
3216  New channel: https://www.youtube.com/channel/U...  
3217  Pokemon emerald but every battle is a 1v2You s...  
3218  Suscríbete: https://Rancho.lnk.to/RHYoutube\r...  
3219  never again ahaha.... CHECK OUT OUR MERCH: htt...  
3220  My Cuphead: The Delicious Last Course Let's Pl...

In [90]:
# Your code here
df.shape

(3221, 15)

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

In [7]:
df.dtypes

country              object
video_id             object
title                object
publishedAt          object
channelId            object
channelTitle         object
categoryId            int64
trending_date        object
tags                 object
view_count            int64
comment_count         int64
thumbnail_link       object
comments_disabled      bool
ratings_disabled       bool
description          object
dtype: object

#### Задание 2 

Ответьте на вопросы:

* Для скольких стран (колонка country) есть информация о видеороликах в таблице (отсутствие страны НЕ учитывается как +1 значение)?
* Сколько есть видеороликов (== записей в таблице) c разрешенными и запрещенными комментариями? (колонка comments_disabled). Под видеороликом считаем 1 запись в таблице.
* Какое максимальное число просмотров у видеороликов, попавших в таблицу? (колонка view_count)

In [ ]:
df['']

In [62]:
# Your code here
df.loc[df['comments_disabled'] == True]['comments_disabled'].count()

np.int64(28)

In [64]:
df.loc[df['comments_disabled'] == False]['comments_disabled'].count()

np.int64(3193)

In [65]:
df['view_count'].max()

np.int64(123034293)

**Ответ**:
1. ВАШ ОТВЕТ ЗДЕСЬ
2. Разрешено 28, запрещено 3193
3. 123034293

#### Задание 3

* Есть ли в данных пропуски? В каких колонках?

In [92]:
# Your code here
df.isna().sum()

country              200
video_id               0
title                  0
publishedAt            0
channelId              0
channelTitle           0
categoryId             0
trending_date          0
tags                   0
view_count             0
comment_count          0
thumbnail_link         0
comments_disabled      0
ratings_disabled       0
description          131
dtype: int64

In [93]:
df.dtypes

country              object
video_id             object
title                object
publishedAt          object
channelId            object
channelTitle         object
categoryId            int64
trending_date        object
tags                 object
view_count            int64
comment_count         int64
thumbnail_link       object
comments_disabled      bool
ratings_disabled       bool
description          object
dtype: object

**Ответ**: country 200, description 131

In [94]:
df.fillna(
    {
        "description": " ",
    },
    inplace=True,
)

In [113]:
df = df.dropna()

Заполните пропуски пустой строкой для текстовых колонок, и удалите строки с пропусками для колонок с категориальными значениями.

**Внимание!** Для проверки корректности выведите кол-во пропусков во всех столбцах и общее кол-во строк в таблице - должна быть 3021 запись. Ошибка при удалении строк может привести к неправильному выполнению следующих заданий.

In [114]:
df.shape

(3021, 15)

In [115]:
assert(df.shape[0] == 3021, "Проверьте правильность выполнения задания.")

<>:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
<>:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
C:\Users\janep\AppData\Local\Temp\ipykernel_13496\1825482647.py:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
  assert(df.shape[0] == 3021, "Проверьте правильность выполнения задания.")


#### Задание 4 

Посмотрите на колонку publishedAt
* Какого она типа?
* Измените тип колонки на datetime. Покажите, что после изменения колонка действительно имеет тип datetime.
* В какой месяц было добавлено больше всего видеороликов?

In [117]:
# Your code here
df['publishedAt'].dtype

dtype('O')

**Ответ**:
1. ВАШ ОТВЕТ ЗДЕСЬ
2. ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 5 

Посмотрите на колонку description.
* Переведите строки в нижний регистр и обрежьте ненужные пробелы справа и слева.
* Замените url-адреса  в описаниях на специальный тэг \<URL\>.
* Выведите полностью description для видеоролика с video_id = 'Oxexirlmdig'.

In [ ]:
pattern = r'http\S+'

In [ ]:
# Your code here

#### Задание 6 
* Какие 5 роликов набрали в Италии (country='IT') наибольшее число просмотров (колонка view_count)? В качестве ответа приведите названия каналов с соответствующими роликами (колонка channelTitle) в порядке убывания числа просмотров полученных роликов.

Для решения воспользуйтесь функцией [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 7 
* Какой видеоролик был опубликован в наибольшем кол-ве стран? В качестве ответа привeдите название видеоролика (колонка title).

Для решения могут помочь функции [groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) и [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 8
* Какое максимальное число просмотров (колонка view_count) у видеороликов, выпущенных во Франции или Италии (в колонке country значениe FR или IT) в июле (колонка publishedAt)?

Необходимо придумать решение в одну строку, используя для выбора строк несколько условий, объединенных с помощью | вместо or, & вместо and и ~ вместо not, а также используя функцию [isin](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html).

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 9 
* Посмотрите на названия каналов из 17 категории (колонка categoryId). Как вы думаете, что это за категория?
* Предположим, что была допущена ошибка, и все видеоролики из 17 категории, относятся к 16 категории. Исправьте нужные значения в колонке categoryId, после чего выведите все уникальные категории.

Обратите внимание, что при выборе полей с помощью оператора индексирования \[\] возвращается копия выбранных значений. При попытке присвоить таким образом новые значения для некоторых полей таблицы, будет выдано предупреждение. Более надежный способ - использовать операторы loc и iloc.

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 10 
Посмотрите на колонку tags.

* Замените строки со значением '[none]' на пустую строку.
* Какой самый популярный тэг? Обратите внимание, что у каждого видеоролика может быть более 1 тэга.

**Подсказка 1**

Чтобы получить правильное значение, не забудьте обработать множество тэгов - убрать пробельные символы по краям и привести строки в нижний регистр.

**Подсказка 2**

Нужный тэг встречается больше 130 раз.

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

# Часть 2. Визуализация данных


### Оформление графиков

Важно, чтобы все графики выглядели опрятно и соответствовали следующим правилам:

- должно быть название (plt.title) графика;
- на графиках должны быть подписаны оси (plt.xlabel, plt.ylabel);
- должны быть подписаны единицы измерения (если это возможно);
- все названия должны быть понятны любому человеку, знакомому с терминологией, без заглядывания в код;
- подписи тиков на осях не должны сливаться как на одной оси, так и между ними;
- если изображено несколько сущностей на одном холсте (например несколько функций), то необходима поясняющая легенда (plt.legend);
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- графики должны быть не супер-микро и не супер-макро по размерам, так, чтобы можно было увидеть все, что нужно.
- при необходимости улучшения наглядности графиков, можно пользоваться логарифмической шкалой по осям x/y.

In [ ]:
import matplotlib
import matplotlib.pyplot as plt

#отображение графиков в ноутбуке
%matplotlib inline

#для четкой прорисовки графиков - осторожно!!! Может съедать много памяти
#закоменьте эту строку, если это случилось
%config InlineBackend.figure_format = 'svg'

### Задание 1

Постройте график (line plot) зависимости кол-ва выпущенных роликов от даты публикации.

Обратите внимание, что один и тот же ролик может быть опубликован в нескольких странах, и ему может соответствовать несколько строк в таблице. На графике необходимо учесть каждый ролик только 1 раз.

In [ ]:
# Your code here

### Задание 2
Постройте диаграмму  (bar plot) кол-ва опубликованных роликов в зависимости от страны.

In [ ]:
# Your code here

### Задание 3 
Постройте график (scatter plot) зависимости кол-ва комментариев для ролика от кол-ва просмотров для роликов во Франции (country='FR'). На графике не нужно учитывать ролики с отключенными комментариями (comments_disabled=True). Какой характер зависмости получился?

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ